File: BI-TJV/semestral/rest-server/src/main/java/wrzecond/service/TjvOrderService.java
package wrzecond.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import wrzecond.order.TjvOrderCreateDTO;
import wrzecond.order.TjvOrderReadDTO;
import wrzecond.entity.TjvOrder;
import wrzecond.entity.TjvTable;
import wrzecond.repository.TjvOrderRepository;

import javax.persistence.EntityNotFoundException;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class TjvOrderService extends TjvService<TjvOrder, TjvOrderReadDTO, TjvOrderCreateDTO> {

    // === PARAMETERS ===
    private final TjvOrderRepository orderRepository;
    private final TjvTableService tableService;

    @Autowired
    public TjvOrderService(TjvOrderRepository orderRepository, TjvTableService service) {
        super (orderRepository);
        this.orderRepository = orderRepository;
        this.tableService = service;
    }

    public List<TjvOrderReadDTO> findByTableID (int tableId) {
        // Table must exist
        TjvTable table = tableService.getEntityByID(tableId);
        if (table == null)
            throw new EntityNotFoundException();

        // Return (correct)
        return orderRepository.findByTableID(table)
            .stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    @Override
    protected TjvOrderReadDTO toDTO (TjvOrder order) {
        return new TjvOrderReadDTO(order.getId(), order.getTable().getId(), order.getDatetime(), order.isCompleted());
    }

    @Override
    protected TjvOrder fromDTO (TjvOrderCreateDTO dto) {
        return new TjvOrder(
            tableService.getEntityByID(dto.getTable()),
            dto.getDatetime(), dto.getCompleted()
        );
    }

    @Override
    protected TjvOrder merge (TjvOrder entity, TjvOrderCreateDTO dto) {
        if (dto.getDatetime() != null)
            entity.setDatetime(dto.getDatetime());
        if (dto.getCompleted() != null)
            entity.setCompleted(dto.getCompleted());

        // Table
        if (dto.getTable() != null)
            entity.setTable(tableService.getEntityByID(dto.getTable()));

        return entity;
    }

}

File: BI-KOT/semestral/rest-server/src/main/kotlin/wrzecond/service/TjvOrderService.kt
package wrzecond.service

import org.springframework.stereotype.Service
import wrzecond.TjvOrderCreateDTO
import wrzecond.TjvOrderReadDTO
import wrzecond.TjvOrderUpdateDTO
import wrzecond.entity.TjvOrder
import wrzecond.repository.TjvOrderRepository

@Service
class TjvOrderService (override val repository: TjvOrderRepository, private val tableService: TjvTableService)
: TjvService<TjvOrder, TjvOrderReadDTO, TjvOrderCreateDTO, TjvOrderUpdateDTO>(repository) {

    fun findByTableID (tableId: Int)
        = repository.findByTableID(tableService.getEntityByID(tableId)).map { it.toDTO() }

    override fun TjvOrder.toDTO() = TjvOrderReadDTO (id, table.id, datetime, completed)
    override fun TjvOrderCreateDTO.toEntity() = TjvOrder (tableService.getEntityByID(table), datetime, completed)
    override fun TjvOrder.merge (dto: TjvOrderUpdateDTO) = TjvOrder (
        dto.table?.let { tableService.getEntityByID(it) } ?: table,
        dto.datetime ?: datetime,
        dto.completed ?: completed,
        id
    )

}